feat/contract-tooling

denalimarsh opened the pull request:
This PR enhances Peggy smart contract interaction by integrating truffle into the project's development environment and usage.

- [x] Importing OpenZeppelin contracts directly from openzeppelin-solidity library
- [x] Local end-to-end testing with truffle: deploy peggy contract, ebrelayer connection, send lock transaction to peggy contract
- [x] Yarn workspaces and user-friendly commands: `yarn peggy:abi`, `yarn peggy:address`, `yarn:peggy lock`, etc.
- [x] Ropsten end-to-end test with Infura and MetaMask
- [x] Updated README with all commands/steps for end-to-end test

This PR has been preliminarily reviewed in the swishlabsco/peggy repo: https://github.com/swishlabsco/peggy/pull/2.

***Lock transaction on local truffle blockchain***
<img width="716" alt="Screen Shot 2019-09-13 at 12 07 09 PM" src="https://user-images.githubusercontent.com/15370712/64888175-22577d00-d61f-11e9-8b53-3a858e566628.png">

***ebrelayer witnesses LogLock event***
<img width="1082" alt="Screen Shot 2019-09-13 at 12 07 39 PM" src="https://user-images.githubusercontent.com/15370712/64888186-27b4c780-d61f-11e9-9d5f-0bc1d944483f.png">

_Future work:_
tmux bash script which executes a local end-to-end test

musnit commented:
@denalimarsh what happened to package.json?

musnit commented:
@denalimarsh can you update circleci to do truffle stuff/some tests as well

denalimarsh commented:
> @denalimarsh package.lock, vendor and package.json should all be in same folder shouldn't they?

Fixed in https://github.com/cosmos/peggy/pull/63/commits/04c4b501422031a88f0b85ffafe41070acda2602

denalimarsh commented:
> @denalimarsh can you update circleci to do truffle stuff/some tests as well

CircleCI for smart contracts is slated for next milestone. I've created an issue for this: https://github.com/cosmos/peggy/issues/66

musnit commented on README.md:
```diff
@@ -112,49 +113,125 @@ ebcli rest-server --trust-node
+
+### Set-up
+
+Create a .env file with environment variable LOCAL_PROVIDER - an example configuration can be found in .env.example. For running the bridge locally, you'll only need the LOCAL_PROVIDER. Environment variables MNEMONIC and INFURA_PROJECT_ID are only required for using the Ropsten testnet (see below).
```
would be nice to add these steps for users to copypase, ie:

```bash
cd testnet-contracts/
yarn cp .env.example .env
```

denalimarsh replied:
https://github.com/cosmos/peggy/pull/63/commits/46b75761ac290b4459c284cd80c6ea4b22659041, https://github.com/cosmos/peggy/pull/63/commits/3aa4a1fea4fa417de6415a6aa9488addc81cc407

fedekunze commented on README.md:
````diff
@@ -112,49 +113,138 @@ ebcli rest-server --trust-node
+cp .env.example .env
+```
+
+For running the bridge locally, you'll only need the LOCAL_PROVIDER environment variables. Environment variables MNEMONIC and INFURA_PROJECT_ID are required for using the Ropsten testnet.
````
```suggestion
For running the bridge locally, you'll only need the `LOCAL_PROVIDER` environment variables. Environment variables `MNEMONIC` and `INFURA_PROJECT_ID` are required for using the Ropsten testnet.
```

denalimarsh replied:
https://github.com/cosmos/peggy/pull/63/commits/b1b83debc06de1ba44ec8f43bc0d72333e1e5c58

fedekunze commented on README.md:
````diff
@@ -112,49 +113,138 @@ ebcli rest-server --trust-node
+
+### Set-up
+
+```
````
```suggestion
```bash
```

denalimarsh replied:
https://github.com/cosmos/peggy/pull/63/commits/b1b83debc06de1ba44ec8f43bc0d72333e1e5c58

fedekunze commented on README.md:
```diff
@@ -112,49 +113,138 @@ ebcli rest-server --trust-node
+
+Further reading:
+
+[MetaMask Mnemonic](https://metamask.zendesk.com/hc/en-us/articles/360015290032-How-to-Reveal-Your-Seed-Phrase)
```
```suggestion
- [MetaMask Mnemonic](https://metamask.zendesk.com/hc/en-us/articles/360015290032-How-to-Reveal-Your-Seed-Phrase)
```

denalimarsh replied:
https://github.com/cosmos/peggy/pull/63/commits/b1b83debc06de1ba44ec8f43bc0d72333e1e5c58

fedekunze commented on README.md:
```diff
@@ -112,49 +113,138 @@ ebcli rest-server --trust-node
+Further reading:
+
+[MetaMask Mnemonic](https://metamask.zendesk.com/hc/en-us/articles/360015290032-How-to-Reveal-Your-Seed-Phrase)
+
```
```suggestion
```

denalimarsh replied:
https://github.com/cosmos/peggy/pull/63/commits/b1b83debc06de1ba44ec8f43bc0d72333e1e5c58

fedekunze commented on README.md:
```diff
@@ -112,49 +113,138 @@ ebcli rest-server --trust-node
+
+[MetaMask Mnemonic](https://metamask.zendesk.com/hc/en-us/articles/360015290032-How-to-Reveal-Your-Seed-Phrase)
+
+[Infura Project ID](https://blog.infura.io/introducing-the-infura-dashboard-8969b7ab94e7)
```
```suggestion
- [Infura Project ID](https://blog.infura.io/introducing-the-infura-dashboard-8969b7ab94e7)
```

denalimarsh replied:
https://github.com/cosmos/peggy/pull/63/commits/b1b83debc06de1ba44ec8f43bc0d72333e1e5c58

fedekunze commented on README.md:
````diff
@@ -112,49 +113,138 @@ ebcli rest-server --trust-node
+
+### Terminal 1: Start local blockchain
+
+```
````
```suggestion
```bash
```

denalimarsh replied:
https://github.com/cosmos/peggy/pull/63/commits/b1b83debc06de1ba44ec8f43bc0d72333e1e5c58

fedekunze commented on README.md:
````diff
@@ -112,49 +113,138 @@ ebcli rest-server --trust-node
+### Terminal 2: Compile and deploy Peggy contract
+
+```
+
````
```suggestion
```

denalimarsh replied:
https://github.com/cosmos/peggy/pull/63/commits/b1b83debc06de1ba44ec8f43bc0d72333e1e5c58

fedekunze commented on README.md:
````diff
@@ -112,49 +113,138 @@ ebcli rest-server --trust-node
+
+### Terminal 2: Compile and deploy Peggy contract
+
+```
````
```suggestion
```bash
```

denalimarsh replied:
https://github.com/cosmos/peggy/pull/63/commits/b1b83debc06de1ba44ec8f43bc0d72333e1e5c58

fedekunze commented on README.md:
````diff
@@ -112,49 +113,138 @@ ebcli rest-server --trust-node
+
+### Terminal 3: Build and start Ethereum Bridge
+
+```
````
```suggestion
```bash
```

denalimarsh replied:
https://github.com/cosmos/peggy/pull/63/commits/b1b83debc06de1ba44ec8f43bc0d72333e1e5c58

fedekunze commented on README.md:
````diff
@@ -112,49 +113,138 @@ ebcli rest-server --trust-node
+
+For automated relaying, there is a relayer service that can be run that will automatically watch and relay events (local web socket and deployed address parameters may vary).
 
 ```
````
```suggestion
```bash
```

denalimarsh replied:
https://github.com/cosmos/peggy/pull/63/commits/b1b83debc06de1ba44ec8f43bc0d72333e1e5c58

fedekunze commented on README.md:
```diff
@@ -112,49 +113,138 @@ ebcli rest-server --trust-node
+# You should see a message like: Started ethereum websocket with provider: [LOCAL_WEB_SOCKET] \ Subscribed to contract events on address: [PEGGY_DEPLOYED_ADDRESS]
+
+# The relayer will now watch the contract on Ropsten and create a claim whenever it detects a lock event.
+
```
```suggestion
```

fedekunze commented on README.md:
````diff
@@ -112,49 +113,138 @@ ebcli rest-server --trust-node
-The relayer will now watch the contract on Ropsten and create a claim whenever it detects a lock event.
+### Using Terminal 2: Send lock transaction to contract
+
+```
````
```suggestion
```bash
```

denalimarsh replied:
https://github.com/cosmos/peggy/pull/63/commits/b1b83debc06de1ba44ec8f43bc0d72333e1e5c58

fedekunze commented on testnet-contracts/.gitignore:
```diff
@@ -1,2 +1,3 @@
 build/
-node_modules/ 
+
```
```suggestion
.env
```

denalimarsh replied:
https://github.com/cosmos/peggy/pull/63/commits/b1b83debc06de1ba44ec8f43bc0d72333e1e5c58

fedekunze commented on testnet-contracts/.gitignore:
```diff
@@ -1,2 +1,3 @@
 build/
-node_modules/ 
+
+.env
```
```suggestion

```

denalimarsh replied:
https://github.com/cosmos/peggy/pull/63/commits/b1b83debc06de1ba44ec8f43bc0d72333e1e5c58

fedekunze commented on testnet-contracts/scripts/getDeployedAddress.js:
```diff
@@ -0,0 +1,44 @@
+   *** Set up
+   ******************************************/
+  require("dotenv").config();
+  let Web3 = require("web3");
```
```suggestion
  const Web3 = require("web3");
```

denalimarsh replied:
https://github.com/cosmos/peggy/pull/63/commits/b1b83debc06de1ba44ec8f43bc0d72333e1e5c58

fedekunze commented on testnet-contracts/scripts/getDeployedAddress.js:
```diff
@@ -0,0 +1,44 @@
+   ******************************************/
+  require("dotenv").config();
+  let Web3 = require("web3");
+  var HDWalletProvider = require("@truffle/hdwallet-provider");
```
```suggestion
  const HDWalletProvider = require("@truffle/hdwallet-provider");
```

denalimarsh replied:
https://github.com/cosmos/peggy/pull/63/commits/b1b83debc06de1ba44ec8f43bc0d72333e1e5c58

fedekunze commented on testnet-contracts/scripts/getDeployedAddress.js:
```diff
@@ -0,0 +1,44 @@
+    provider = new Web3.providers.HttpProvider(process.env.LOCAL_PROVIDER);
+  }
+
+  var web3 = new Web3(provider);
```
```suggestion
  let web3 = new Web3(provider);
```

denalimarsh replied:
https://github.com/cosmos/peggy/pull/63/commits/b1b83debc06de1ba44ec8f43bc0d72333e1e5c58

fedekunze commented on testnet-contracts/scripts/getDeployedAddress.js:
```diff
@@ -0,0 +1,44 @@
+
+  // Contract abstraction
+  const truffleContract = require("truffle-contract");
+  let contract = truffleContract(require("../build/contracts/Peggy.json"));
```
```suggestion
  const contract = truffleContract(require("../build/contracts/Peggy.json"));
```

denalimarsh replied:
https://github.com/cosmos/peggy/pull/63/commits/b1b83debc06de1ba44ec8f43bc0d72333e1e5c58

fedekunze commented on testnet-contracts/scripts/getDeployedAddress.js:
```diff
@@ -0,0 +1,44 @@
+    return instance.address;
+  });
+
+  return console.log("Deployed contract address: ", address);
```
why return a console log?

denalimarsh replied:
https://github.com/cosmos/peggy/pull/63/commits/b1b83debc06de1ba44ec8f43bc0d72333e1e5c58

fedekunze commented on testnet-contracts/truffle-config.js:
```diff
@@ -0,0 +1,39 @@
+      host: "localhost",
+      port: 7545,
+      network_id: "*",
+      gas: 6721975,
```
comments on these values?

denalimarsh replied:
https://github.com/cosmos/peggy/pull/63/commits/b1b83debc06de1ba44ec8f43bc0d72333e1e5c58

fedekunze commented on testnet-contracts/scripts/sendLockTx.js:
```diff
@@ -0,0 +1,126 @@
+   *** truffle exec lacks support for dynamic command line arguments:
+   *** https://github.com/trufflesuite/truffle/issues/889#issuecomment-522581580
+   ******************************************/
+  if (NETWORK_ROPSTEN && DEFAULT_PARAMS) {
```
a `switch` statement here would improve readability for the outer conditions

denalimarsh replied:
Planning on refactoring to node scripts instead of using truffle exec due to the difficulties parsing/validating command line arguments and also its latency (script runs for ~25 seconds despite logging the LogLock information right away). This section will be deprecated shortly in favor of simpler input validation in the new script.

fedekunze commented on testnet-contracts/scripts/sendLockTx.js:
```diff
@@ -0,0 +1,126 @@
+  /*******************************************
+   *** Set up
+   ******************************************/
+  let Web3 = require("web3");
```
```suggestion
  const Web3 = require("web3");
```

denalimarsh replied:
https://github.com/cosmos/peggy/pull/63/commits/b1b83debc06de1ba44ec8f43bc0d72333e1e5c58

fedekunze commented on testnet-contracts/scripts/sendLockTx.js:
```diff
@@ -0,0 +1,126 @@
+   *** Set up
+   ******************************************/
+  let Web3 = require("web3");
+  var HDWalletProvider = require("@truffle/hdwallet-provider");
```
```suggestion
  const HDWalletProvider = require("@truffle/hdwallet-provider");
```

denalimarsh replied:
https://github.com/cosmos/peggy/pull/63/commits/b1b83debc06de1ba44ec8f43bc0d72333e1e5c58

fedekunze commented on testnet-contracts/scripts/sendLockTx.js:
```diff
@@ -0,0 +1,126 @@
+
+  // Contract abstraction
+  const truffleContract = require("truffle-contract");
+  let contract = truffleContract(require("../build/contracts/Peggy.json"));
```
```suggestion
  const contract = truffleContract(require("../build/contracts/Peggy.json"));
```

denalimarsh replied:
https://github.com/cosmos/peggy/pull/63/commits/b1b83debc06de1ba44ec8f43bc0d72333e1e5c58

denalimarsh commented:
Thanks for the review @fedekunze, I've pushed the requested changes. I've also rebased this feature to the new cosmos sdk update in https://github.com/cosmos/peggy/pull/65 and simplified the yarn workspace cmds.

fedekunze commented:
@denalimarsh can you resolve the conflicts with master ? 🙏

denalimarsh commented:
> @denalimarsh can you resolve the conflicts with master ? 🙏

@fedekunze  conflicts resolved in latest push

fedekunze commented on app/app.go~HEAD:
```diff
@@ -0,0 +1,245 @@
+package app
```
the file shouldn't be called `app.go~HEAD` but `app.go`

denalimarsh replied:
https://github.com/cosmos/peggy/pull/63/commits/1e4c1e7bd6febbb1bc68144caed1529a3e2754f1

fedekunze commented on cmd/ebrelayer/contract/contract_test.go:
```diff
@@ -12,11 +9,10 @@ import (
 // TestLoadABI : test that contract containing named event is successfully loaded
 func TestLoadABI(t *testing.T) {
 
+	const AbiPath = "/src/github.com/cosmos/peggy/cmd/ebrelayer/contract/abi/Peggy.abi"
```
^^

denalimarsh replied:
https://github.com/cosmos/peggy/pull/63/commits/6115766bf182fec925f334142cbc8bffa557c641

fedekunze commented on cmd/ebrelayer/relayer/network_test.go:
```diff
@@ -18,7 +18,7 @@ func TestIsWebsocketURL(t *testing.T) {
 // TestSetupWebsocketEthClient : test initialization of Ethereum websocket
 func TestSetupWebsocketEthClient(t *testing.T) {
-	_, err := SetupWebsocketEthClient(Client)
+	_, err := SetupWebsocketEthClient(Client, "testing")
```
^^

denalimarsh replied:
https://github.com/cosmos/peggy/pull/63/commits/7ad27ff17dff775a6e7ae667315b88b7b6e981cc

denalimarsh commented on x/oracle/module.go:
```diff
@@ -0,0 +1,131 @@
+}
+
+// NewAppModule creates a new AppModule object
+func NewAppModule(keeper Keeper) AppModule {
```
https://github.com/cosmos/peggy/pull/63/commits/3532e86a60fefaf6e2f43eab5e6bccd50a272c97

denalimarsh commented on app/app.go:
```diff
@@ -0,0 +1,245 @@
+
+// NewEthereumBridgeApp is a constructor function for EthereumBridgeApp
+func NewEthereumBridgeApp(logger log.Logger, db dbm.DB,
+	baseAppOptions ...func(*bam.BaseApp)) *EthereumBridgeApp {
```
https://github.com/cosmos/peggy/pull/63/commits/3532e86a60fefaf6e2f43eab5e6bccd50a272c97

denalimarsh commented on x/ethbridge/module.go:
```diff
@@ -0,0 +1,145 @@
+}
+
+// NewAppModule creates a new AppModule object
+func NewAppModule(oracleKeeper oracle.Keeper, bankKeeper bank.Keeper, codespace sdk.CodespaceType, cdc *codec.Codec) AppModule {
```
https://github.com/cosmos/peggy/pull/63/commits/3532e86a60fefaf6e2f43eab5e6bccd50a272c97

fedekunze commented on testnet-contracts/contracts/Processor.sol:
```diff
@@ -1,7 +1,8 @@
 pragma solidity ^0.5.0;
 
-import "openzeppelin-solidity/contracts/math/SafeMath.sol";
```
question: why do we import using `node_modules` now?
